### PR TITLE
[Grappler] Support unknown shapes in SimplifyReduction

### PR DESCRIPTION
**Pull request by benbarsdell**
Allows reductions over unknown shapes to be replaced with Identity when the set of reduction indices is empty (which is common in the gradient graph). Previously this only worked if the input shape was known.

Extends the tests to include this case.

This can have a big impact on the subsequent auto_mixed_precision pass, where the presence of reductions prevents optimizations.

FYI @reedwm

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/constant_folding.cc [hunk 1]
```diff
@@ -2382,20 +2382,34 @@ bool ConstantFolding::SimplifySwitch(GraphDef* optimized_graph, NodeDef* node) {
...
   NodeDef* reductions_indices = node_map_->GetNode(node.input(1));
-  if (!IsReallyConstant(*reductions_indices)) {
+  if (!IsReallyConstant(*reductions_indices) ||
+      !reductions_indices->attr().count("value")) {
     return false;
   }
+  const TensorProto& reduction_indices_tensor =
+      reductions_indices->attr().at("value").tensor();
+  *indices_is_empty = false;
+  for (const auto& dim : reduction_indices_tensor.tensor_shape().dim()) {
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/constant_folding.cc]: Would it be better to use TensorShapeUtils here?

Comment (benbarsdell) [reply tensorflow/core/grappler/optimizers/constant_folding.cc]: Thanks for the suggestion. I've replaced it with `TensorShape(reduction_indices_shape).num_elements()`.

File: tensorflow/core/grappler/optimizers/constant_folding.cc [hunk 2]
```diff
@@ -2460,9 +2468,33 @@ bool ConstantFolding::IsReductionSimplifiableToIdentity(
   return simplifiable;
 }
 
+bool ConstantFolding::ReplaceReductionWithIdentity(NodeDef* node) const {
+  // Replace the reduction node with an identity node, that can be further
+  // optimized by the model pruner.
```
Comment (rmlarsen) [inline tensorflow/core/grappler/optimizers/constant_folding.cc]: maybe remove "by the model pruner". There are various optimizers that target Identity, most notably in the dependency optimizer.

File: tensorflow/core/grappler/optimizers/constant_folding.cc [hunk 3]
```diff
@@ -2460,9 +2468,33 @@ bool ConstantFolding::IsReductionSimplifiableToIdentity(
...
+  DataType output_type;
+  if (node->attr().count("T") != 0) {
+    output_type = node->attr().at("T").type();
+  } else {
+    // This is an 'any' or 'all' reduction. The output is always boolean.
```
Comment (rmlarsen) [inline tensorflow/core/grappler/optimizers/constant_folding.cc]: Either check that it is indeed such a reduction or return false. Somebody might add new reduction types in the future. Alternatively state in the method-level docstring that this is a precondition that the caller should guarantee.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by rmlarsen** [comment]
@benbarsdell thanks for this improvement. I only have a couple of nits.

File: tensorflow/core/grappler/optimizers/constant_folding.cc (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/core/grappler/optimizers/constant_folding.cc]: Done.

File: tensorflow/core/grappler/optimizers/constant_folding.cc (hunk 3 above)
Comment (benbarsdell) [reply tensorflow/core/grappler/optimizers/constant_folding.cc]: Done. (I'll note that it's not my code; I only moved it :) )

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@benbarsdell thanks for the quick fix. Sorry it took a while to get to reviewing this.

**Pull request merged**